File: Python/String/M-5. Longest Palindromic Substring.py
class Solution:
    @classmethod
    def longestPalindrome(self, s: str) -> str:
        ans = ""
        for i in range(len(s)):
            # odd palindrome
            l, r = i - 1, i + 1
            palindrome = self.get_palindrome(s, l, r)
            ans = palindrome if len(palindrome) > len(ans) else ans

            # even palindrome
            l, r = i - 1, i
            palindrome = self.get_palindrome(s, l, r)
            ans = palindrome if len(palindrome) > len(ans) else ans

        return ans

    @classmethod
    def get_palindrome(self, s: str, l: int, r: int):
        while l >= 0 and r < len(s) and s[l] == s[r]:
            l -= 1
            r += 1

        return s[l+1: r]


test_data = [
    {
        "s": 'babad',
        "ans": 'bab',
    },
    {
        "s": 'cbbd',
        "ans": 'bb',
    },
    {"s": 'aaaa', "ans": 'aaaa'},
]

for data in test_data:
    print(data["ans"] == Solution.longestPalindrome(data["s"]))

File: Javascript/String/M-5. Longest Palindromic Substring.js
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function (s) {
  let ans = '';
  for (let i = 0; i < s.length; i++) {
    // odd palindrome
    let l = i - 1;
    let r = i + 1;
    let palindrome = getPalindrome(s, l, r);
    if (ans.length < palindrome.length) ans = palindrome;

    // even palindrome
    l = i - 1;
    r = i;
    palindrome = getPalindrome(s, l, r);
    if (ans.length < palindrome.length) ans = palindrome;
  }
  return ans;
};

const getPalindrome = (s, l, r) => {
  while (l >= 0 && r < s.length && s[l] === s[r]) {
    l--;
    r++;
  }
  return s.substring(l + 1, r);
};
const testData = [
  {
    s: 'babad',
    ans: 'bab',
  },
  {
    s: 'cbbd',
    ans: 'bb',
  },
  { s: 'aaaa', ans: 'aaaa' },
];

for (const data of testData) {
  console.log(data.ans === longestPalindrome(data.s));
}
